src/hotspot/share/opto/type.cpp

```diff
@@ -5637,81 +5775,117 @@ const Type    *TypeInstKlassPtr::xmeet( const Type *t ) const {
+  static_assert(std::is_base_of<T2, T1>::value, "");
+  if (!this_one->is_loaded() || !other->is_loaded()) {
     return false;
   }
-  if (!other->isa_instklassptr()) {
+  if (!this_one->is_instance_type(other)) {
     return false;
   }
-  return _klass->equals(other->_klass);
+  return this_one->_klass->equals(other->_klass) && this_one->_interfaces.eq(other->_interfaces);
 }
 
-bool TypeInstKlassPtr::maybe_java_subtype_of_helper(const TypeKlassPtr* other, bool this_exact, bool other_exact) const {
-  if (!is_loaded() || !other->is_loaded()) {
-    return true;
-  }
+bool TypeInstKlassPtr::is_same_java_type_as_helper(const TypeKlassPtr* other) const {
+  return TypePtr::is_same_java_type_as_helper_for_instance(this, other);
+}
 
-  if (other->isa_aryklassptr()) {
-    return !this_exact && (_klass->equals(ciEnv::current()->Object_klass()) || _klass->is_interface());
+template <class T1, class T2> bool TypePtr::maybe_java_subtype_of_helper_for_instance(const T1* this_one, const T2* other, bool this_exact, bool other_exact) {
+  static_assert(std::is_base_of<T2, T1>::value, "");
+  if (!this_one->is_loaded() || !other->is_loaded()) {
+    return true;
   }
 
-  if ((_klass->is_interface() && !this_exact) || (other->klass()->is_interface() /*&& !other_exact*/)) {
-    return true;
+  if (this_one->is_array_type(other)) {
+    return !this_exact && this_one->_klass->equals(ciEnv::current()->Object_klass())  && this_one->_interfaces.intersection_with(other->_interfaces).eq(this_one->_interfaces);
   }
 
-  assert(other->isa_instklassptr(), "unsupported");
+  assert(this_one->is_instance_type(other), "unsupported");
 
   if (this_exact && other_exact) {
-    return is_java_subtype_of(other);
+    return this_one->is_java_subtype_of(other);
   }
 
-  if (!_klass->is_subtype_of(other->_klass) && !other->_klass->is_subtype_of(_klass)) {
+  if (!this_one->_klass->is_subtype_of(other->_klass) && !other->_klass->is_subtype_of(this_one->_klass)) {
     return false;
   }
 
   if (this_exact) {
-    return _klass->is_subtype_of(other->_klass);
+    return this_one->_klass->is_subtype_of(other->_klass) && this_one->_interfaces.intersection_with(other->_interfaces).eq(other->_interfaces);
   }
 
   return true;
 }
 
+bool TypeInstKlassPtr::maybe_java_subtype_of_helper(const TypeKlassPtr* other, bool this_exact, bool other_exact) const {
+  return TypePtr::maybe_java_subtype_of_helper_for_instance(this, other, this_exact, other_exact);
+}
+
+const TypeKlassPtr* TypeInstKlassPtr::try_improve() const {
+  ciKlass* k = klass();
+  Compile* C = Compile::current();
+  Dependencies* deps = C->dependencies();
+  assert((deps != NULL) == (C->method() != NULL && C->method()->code_size() > 0), "sanity");
+  TypePtr::InterfaceSet interfaces = _interfaces;
+  if (k->is_loaded()) {
+    ciInstanceKlass* ik = k->as_instance_klass();
+    bool klass_is_exact = ik->is_final();
+    if (!klass_is_exact &&
+        deps != NULL && UseUniqueSubclasses) {
```

Comment:
Please, put `UseUniqueSubclasses` guard at the top of the method.

Reply:
Ok.